cmd/otel-allocator/prehooktargetfilter/relabel_config_filter.go

```diff
@@ -0,0 +1,101 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prehooktargetfilter
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/relabel"
+
+	"github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/allocation"
+)
+
+type RelabelConfigTargetFilter struct {
+	m           sync.RWMutex
+	log         logr.Logger
+	targetItems map[string]*allocation.TargetItem
+	allocator   allocation.Allocator
+}
+
+func NewRelabelConfigTargetFilter(log logr.Logger, allocator allocation.Allocator) AllocatorPrehook {
+	return &RelabelConfigTargetFilter{
+		log:         log,
+		allocator:   allocator,
+		targetItems: make(map[string]*allocation.TargetItem),
+	}
+}
+
+func (tf *RelabelConfigTargetFilter) SetTargets(targets map[string]*allocation.TargetItem) {
+	numRemainingTargets := 0
+	for jobName, tItem := range targets {
+		keepTarget := true
+		for _, cfg := range tItem.RelabelConfigs {
+			if tf.IsDropTarget(tItem.Label, cfg) {
+				keepTarget = false
+				break
+			}
+		}
+
+		if keepTarget {
+			tf.targetItems[jobName] = tItem
+			numRemainingTargets++
+		}
+	}
+
+	tf.allocator.SetTargets(tf.targetItems)
+	tf.log.Info(fmt.Sprintf("Relabel filtering completed. Keeping %d target(s) out of %d total targets", numRemainingTargets, len(targets)))
+}
+
+// TargetItems returns a shallow copy of the targetItems map.
+func (tf *RelabelConfigTargetFilter) TargetItems() map[string]*allocation.TargetItem {
+	tf.m.RLock()
+	defer tf.m.RUnlock()
+	targetItemsCopy := make(map[string]*allocation.TargetItem)
+	for k, v := range tf.targetItems {
+		targetItemsCopy[k] = v
+	}
+	return targetItemsCopy
+}
+
+// Goal of this function is to determine whether a given target should
+// be dropped or not - function should be called for each item in the relabel_config.
+func (tf *RelabelConfigTargetFilter) IsDropTarget(lset model.LabelSet, cfg *relabel.Config) bool {
+	values := make([]string, 0, len(cfg.SourceLabels))
+	for _, ln := range cfg.SourceLabels {
+		if val, ok := lset[ln]; ok {
+			values = append(values, string(val))
+		} else {
+			tf.log.Info(fmt.Sprintf("label %v not found in lset, skipping..", ln))
+			return false
+		}
+	}
+	val := strings.Join(values, cfg.Separator)
+
+	if cfg.Action == "drop" {
+		if cfg.Regex.MatchString(val) {
+			return true
+		}
+	} else if cfg.Action == "keep" {
+		if !cfg.Regex.MatchString(val) {
+			return true
+		}
+	}
```

Comment:
We should be able to re-use Prometheus' logic for dropping here. [This is described here](https://github.com/prometheus/prometheus/blob/main/model/relabel/relabel.go#L191-L204)

Reply:
Can we just call that function?

```go
return relabel.Process(lset, cfg) == nil
```

Should a non-nil return be used?

Reply:
I now refactored to use `relabel.Process` as you both suggested! A non-nil return is used because some relabel_config might apply `keep` or `drop` on a source_label that was created by a prior `relabel` action. (i.e. without the non-nil return, targets could be mistakenly dropped because they couldn't find a source label).